cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,219 @@
+package events
+
+import (
+	"io"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/xdr"
+)
+
+// Cursor represents the position of an event within the sequence of all
+// Soroban events. Soroban events are sorted by ledger sequence, transaction
+// index, operation index, and event index.
+type Cursor struct {
+	// Ledger is the sequence of the ledger which emitted the event.
+	Ledger uint32
+	// Tx is the index of the transaction within the ledger which emitted the event.
+	Tx uint32
+	// Op is the index of the operation within the transaction which emitted the event.
+	Op uint32
+	// Event is the index of the event within in the operation which emitted the event.
+	Event uint32
+}
+
+var (
+	// MinCursor is the smallest possible cursor
+	MinCursor = Cursor{}
+	// MaxCursor is the largest possible cursor
+	MaxCursor = Cursor{
+		Ledger: math.MaxUint32,
+		Tx:     math.MaxUint32,
+		Op:     math.MaxUint32,
+		Event:  math.MaxUint32,
+	}
+)
+
+type event struct {
+	contents xdr.ContractEvent
+	cursor   Cursor
+}
+
+func cmp(a, b uint32) int {
+	if a < b {
+		return -1
+	}
+	if a > b {
+		return 1
+	}
+	return 0
+}
+
+// Cmp compares two cursors.
+// 0 is returned if the c is equal to other.
+// 1 is returned if c is greater than other.
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			return cmp(c.Event, other.Event)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
```

Comment:
I think that there is some misaligment here between the retention window ( which is measured in ledgers) and the events array. Also, it looks like we would store ( in-memory ), redundant information.

It looks to me like the following would be a more efficient memory storage:
```golang
type MemoryStore struct {
        lock            sync.RWMutex
	eventLedgers          []eventLedger
        start            uint32 // starting index into the eventLedgers circular buffer.
}

type eventLedger struct {
   ledgerSeq int
   contractEvents []contractEvent
}
type contractEvent struct {
   xdr.ContractEvent
   // Tx is the index of the transaction within the ledger which emitted the event.
	Tx uint32
	// Op is the index of the operation within the transaction which emitted the event.
	Op uint32
	// Event is the index of the event within in the operation which emitted the event.
	Event uint32
}
```

when `Scan` is called, we could dynamically construct the needed data.
The MemoryStore.eventLedgers would be pre-initialize with the desired retention window.

The goal in the above schema is that there is no copy/realloc of any type during a new ledger.

Reply:
I agree, this makes is much easier to deallocate the the events in a ledger.

It may make sense to extend this hierarchy and and separate the events for each transaction, operation in subslices

```go
type MemoryStore struct {
        lock            sync.RWMutex
	contractEvent    [][][]xdr.ContractEvent
        start           uint32 // starting index into the eventLedgers circular buffer.
        startLedger uint32 // ledger Sequence for the starting index
}
```

This comes with a few advantages and disadvantages

Advantages:
1. You can preallocate most (if not all) the memory, since the maximum transactions per ledger and maximum operations per transactions are known (I don't know if that's the case for events pero operation, but if that's the case, we can also preallocate those)
2. You can use native indexing and subslicing instead of trimming etc ...

Disadvantages:
We need to watch cache behavior being worse due to iterating through subslices